botorch/models/utils/inducing_point_allocators.py

```diff
@@ -0,0 +1,388 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+References
+
+.. [burt2020svgp]
+    David R. Burt and Carl Edward Rasmussen and Mark van der Wilk,
+    Convergence of Sparse Variational Inference in Gaussian Process Regression,
+    Journal of Machine Learning Research, 2020,
+    http://jmlr.org/papers/v21/19-1015.html.
+
+.. [chen2018dpp]
+    Laming Chen and Guoxin Zhang and Hanning Zhou, Fast greedy MAP inference
+    for determinantal point process to improve recommendation diversity,
+    Proceedings of the 32nd International Conference on Neural Information
+    Processing Systems, 2018, https://arxiv.org/abs/1709.05135.
+
+.. [moss2023ipa]
+    Henry B. Moss and Sebastian W. Ober and Victor Picheny,
+    Inducing Point Allocation for Sparse Gaussian Processes
+    in High-Throughput Bayesian Optimization,Proceedings of
+    the 25th International Conference on Artificial Intelligence
+    and Statistics, 2023, https://arxiv.org/pdf/2301.10123.pdf.
+
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable, Union
+
+import torch
+from botorch.models.model import Model
+
+from botorch.utils.probability.utils import ndtr as Phi, phi
+from gpytorch.module import Module
+from linear_operator.operators import LinearOperator
+from torch import Tensor
+
+NEG_INF = -(torch.tensor(float("inf")))
+
+
+class InducingPointAllocator(ABC):
+    r"""
+    This class provides functionality to initialize the inducing point locations
+    of an inducing point-based model, e.g. a `SingleTaskVariationalGP`.
+    """
+
+    @abstractmethod
+    def allocate_inducing_points(
+        inputs: Tensor,
+        covar_module: Module,
+        num_inducing: int,
+        input_batch_shape: torch.Size,
+    ) -> Tensor:
+        """
+        Initialize the `num_inducing` inducing point locations according to a
+        specific initialization strategy.
+
+        Args:
+            inputs: A (*batch_shape, n, d)-dim input data tensor.
+            covar_module: GPyTorch Module returning a LinearOperator kernel matrix.
+            num_inducing: The maximun number (m) of inducing points (m <= n).
+            input_batch_shape: The non-task-related batch shape.
+
+        Returns:
+            A (*batch_shape, m, d)-dim tensor of inducing point locations.
+        """
+
+        pass
+
+    def _allocate_inducing_points(
+        self,
+        inputs: Tensor,
+        covar_module: Module,
+        num_inducing: int,
+        input_batch_shape: torch.Size,
+        quality_function: Callable[[Tensor], Tensor],
+    ) -> Tensor:
+        r"""
+        Private method to allow inducing point allocators to support
+        multi-task models and models with batched inputs.
```

Comment:
It looks like the only thing that happens in the `allocate_inducing_points` functions of the various classes is to instantiate the right quality function. Howe about just making an abstract `_get_quality_function` method that does what its name suggests and just call this as the first thing in `allocate_inducing_points` (which then doesn't need to be abstract anymore)?

Reply:
Great idea! I've updated it and its much much neater